sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala

```diff
@@ -379,14 +366,29 @@ class ParquetFileFormat
           null)
 
       val sharedConf = broadcastedHadoopConf.value.value
+
+      val fileMetaData =
+        ParquetFileReader.readFooter(sharedConf, fileSplit.getPath, SKIP_ROW_GROUPS).getFileMetaData
```

Comment:
will we read footer again in the parquet reader?

Reply:
We should *always* read the footer and adjust the filters to match. In our version we use the following to handle situations where the column names have changed case or where Hive has returned a different case:

```
      // Try to push down filters when filter push-down is enabled.
      val pushed = if (pushdownEnabled) {
        // read the file schema to create Parquet filters that match case
        val fileReader = ParquetFileReader.open(conf, fileSplit.getPath)
        val fileSchema = try {
          new ParquetToSparkSchemaConverter(conf).convert(
            ParquetReadSupport.clipParquetSchema(
              fileReader.getFileMetaData.getSchema, requiredSchema))
        } finally {
          fileReader.close()
        }

        filters
            // Collects all converted Parquet filter predicates. Notice that not all predicates can
            // be converted (`ParquetFilters.createFilter` returns an `Option`). That's why a
            // `flatMap` is used here.
            .flatMap(ParquetFilters.createFilter(fileSchema, _))
            .reduceOption(FilterApi.and)
      } else {
        None
      }
```

That's pretty much the same thing as here. In addition, if columns can be renamed within a Parquet file then you need to push filters for the names used in the file's schema. I don't think that's a problem here because I don't know of a way to rename columns in a Spark Parquet table. (Iceberg uses field IDs to do this.)